### PR TITLE
feat: add `ImplementsDecl` AST node

### LINKED ISSUE #3184: Assigning a value to the `implements` statement is not prevented
### Version Information

* vyper Version (output of `vyper --version`): 0.3.8+commit.e60b021b
* OS: OSX
* Python Version (output of `python --version`): 3.8.0

### What's your issue about?

The `implements` directive is currently translated to an `AnnAssign` AST node with the name `"implements"`. Since no check regarding the value of such AST node is done, it is possible to assign any value to it. Note that since the `implements`'s semantics are used, the value seems to be essentially ignored by the compiler.

As an example, the following contract compiles.

```Vyper
interface A:
    def test(): nonpayable

implements: A = 12

@external
def test():
    pass
```

### How can it be fixed?

Checking that the `value` field of the node is `None` should solve the issue.

### PR DESCRIPTION
**Pull request by tserg**
### What I did

Add `ImplementsDecl` node. Fix #3184.

### How I did it

Add `ImplementsDecl` node by checking `AnnAssign` nodes in `get_node` during AST construction.

### How to verify it

See new test.

### Commit message

```
feat: add `ImplementsDecl` AST node

If a value is assigned for an `implements` statement,
an exception is thrown.
```

### Description for the changelog

Add `ImplementsDecl` AST node, and guard against value assignment for `implements` statement.

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Comment by tserg** [comment]
Note: I think L128 is now redundant, and `vy_ast.VariableDecl` should be added after L121.
https://github.com/vyperlang/vyper/blob/9bdc83a8fbe4f2eaf25c94892daa9adaf960c5c1/vyper/ast/signatures/interface.py#L108-L130

**Review by charles-cooper — COMMENTED** [review]
Approach looks good. Should have a test that `implements: X = Y` is blocked.

File: vyper/ast/nodes.py [hunk 1]
```diff
@@ -1377,6 +1380,16 @@ class ImportFrom(_Import):
     __slots__ = ("level", "module")
 
 
+class ImplementsDecl(Stmt):
+    __slots__ = ("target", "annotation", "simple")
```
Comment (charles-cooper) [inline vyper/ast/nodes.py:1384]: do we allow non-simple targets?

Comment (tserg) [reply vyper/ast/nodes.py:1384]: hm nope, will make the change and add comments.

File: vyper/ast/grammar.lark [hunk 2]
```diff
@@ -26,7 +26,8 @@ WILDCARD: "*"
 _import_name: NAME
 // NOTE: Don't use DOT here, just a separator
 _import_path: (_import_name ".")* _import_name
-import_alias: _AS NAME
+_import_alias_name: NAME
```
Comment (charles-cooper) [inline vyper/ast/grammar.lark:29]: why is this necessary?

File: vyper/ast/grammar.lark [hunk 3]
```diff
@@ -101,6 +102,8 @@ _INTERFACE_DECL: "interface"
 mutability: NAME
 interface_function: function_sig ":" mutability
 interface_def: _INTERFACE_DECL NAME ":" _NEWLINE _INDENT ( interface_function _NEWLINE)+ _DEDENT
+_IMPLEMENTS_DECL: "implements"
+implements_def: _IMPLEMENTS_DECL ":" ( _import_name | _import_alias_name )
```
Comment (charles-cooper) [inline vyper/ast/grammar.lark:106]: why not just `_IMPLEMENTS_DECL ":" NAME`?

Comment (tserg) [reply vyper/ast/grammar.lark:106]: I thought the RHS of `_IMPLEMENTS_DECL ":" NAME` should be limited to the identifier(s) that have been imported into the namespace, which would be (1) `_import_name`; and (2) the `NAME` in `import_alias: _AS NAME`. 

I think `_IMPLEMENTS_DECL ":" NAME` is more general, but it could also work since the typechecker would catch an identifier that has not been imported.

File: vyper/ast/grammar.lark (hunk 2 above)
Comment (tserg) [reply vyper/ast/grammar.lark:29]: I have included the explanation in my reply to the next comment.

File: vyper/ast/grammar.lark (hunk 3 above)
Comment (charles-cooper) [reply vyper/ast/grammar.lark:106]: well, the LHS is effectively just an alias for the RHS, for readability. it is not like the parser is going to check that the identifiers have been included in import statements, that happens during semantic analysis anyways.

Comment (tserg) [reply vyper/ast/grammar.lark:106]: oh okay got it, will make the change!

**Review by charles-cooper — APPROVED** [review]
thanks, nice work
